File: core/src/main/java/pcb/uwu/units/quantity/TemperatureUnit.java
package pcb.uwu.units.quantity;

import pcb.uwu.core.BaseUnit;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.UnitCounter;

import java.util.function.Function;

public abstract class TemperatureUnit extends BaseUnit {

	private final UnitCounter unitCounter;

	public TemperatureUnit(String symbol, String singularName, String pluralName,
						   Function<BigDecimalAmount, BigDecimalAmount> translationToCanonical,
						   Function<BigDecimalAmount, BigDecimalAmount> translationFromCanonical) {

		super(5, symbol, singularName, pluralName, translationToCanonical, translationFromCanonical);

		this.unitCounter = new UnitCounter(this);
	}

	@Override
	public Class<? extends BaseUnit> getBaseUnitType() {
		return TemperatureUnit.class;
	}

	@Override
	public UnitCounter getUnitCounter() {
		return unitCounter;
	}
}

File: core/src/main/java/pcb/uwu/unit/base/TemperatureUnit.kt
package pcb.uwu.unit.base

import pcb.uwu.core.Amount
import pcb.uwu.core.BaseUnit
import pcb.uwu.core.UnitCounter
import java.util.function.Function

open class TemperatureUnit(symbol: String,
                           singularName: String,
                           pluralName: String,
                           translationToCanonical: Function<Amount, Amount>,
                           translationFromCanonical: Function<Amount, Amount>)
    : BaseUnit(precedence = 5,
               symbol = symbol,
               singularName = singularName,
               pluralName = pluralName,
               translationToCanonical = translationToCanonical,
               translationFromCanonical = translationFromCanonical)
{
    override val unitCounter
        get() = UnitCounter(this)
    override val baseUnitType = TemperatureUnit::class.java
}